plugins/device/device.proto

```diff
@@ -0,0 +1,108 @@
+syntax = "proto3";
+package hashicorp.nomad.plugins.device;
+
+import "google/protobuf/empty.proto";
+
+// DevicePlugin is the API exposed by device plugins
+service DevicePlugin {
+  // FingerprintDevices allows the device plugin to return a set of   
+  // detected devices and provide a mechanism to update the state of
+  // the device.
+  rpc FingerprintDevices(google.protobuf.Empty) returns (stream DetectedDevices) {}
+
+  // Allocate is called by the client before starting an allocation
+  // that requires access to the plugin’s devices. The plugin can use
+  // this to run any setup steps and provides the mounting details to 
+  // the Nomad client
+  rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+
+// DetectedDevices is the set of devices that the device plugin has 
+// detected and is exposing
+message DetectedDevices {
+  // device_name is the name of the device. It should be scoped by 
+  // vendor  (“nvidia/gpu”, “intel/gpu”, “amd/gpu”, …)
+  string device_name = 1;
+  
+  // devices is the set of devices detected by the plugin.
+  repeated DetectedDevice devices = 2;
+
+  // node_attributes allows adding node attributes to be used for 
+  // constraints or affinities.
+  map<string, string> node_attributes = 3;
+}
+
+// DetectedDevice is a single detected device.
+message DetectedDevice {
+  // ID is the ID of the device. This ID is used during 
+  // allocation.
+  string ID = 1;
+
+  // Health of the device.
+  bool healthy = 2;
+  
+  // health_description allows the device plugin to optionally 
+  // annotate the health field with a human readable reason.
+  string health_description = 3;
+
+  // pci_bus_id is the PCI bus ID for the device. If reported, it
+  // allows Nomad to make NUMA aware optimizations.
+  string pci_bus_id = 4; 
+}
+
+// AllocateRequest is used to ask the device driver for information on 
+// how to allocate the requested devices.
+message AllocateRequest {
+  // device_ids are the requested devices.
+  repeated string device_ids = 1;
+}
+
+// AllocateResponse informs Nomad how to expose the requested devices
+// to the the task.
```

Comment:
`Allocation` means something completely different on the server side, wondering if we could come up with a better name that matches what this is for. How about `DeviceMountRequest`, `DeviceMountResponse` and `ContainerMount` instead of `AllocateRequest`, `AllocateResponse` and `ContainerAllocation`?

Reply:
I agree, this tripped me up reading it. I expected something like:
```
rpc AllocateDevice(AllocateDeviceRequest) returns (AllocateDeviceResponse)
or 
rpc ReserveDevice(ReserveDeviceRequest) returns (ReserveDeviceResponse)
```

Reply:
Sure, if it is a source of confusion I will change it

Reply:
Going to do Reserve and not ReserveDevice since it is already on the device interface